### PR TITLE
Remove redundant unwrap in `ScalarValue::new_primitive`, return a `Result` 

### LINKED ISSUE #3313: Review use of panic in `datafusion-common` crate
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
```
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/from_slice.rs":68             length_so_far += OffsetSize::from_usize(s.len()).unwrap();
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/pyarrow.rs":70         self.to_pyarrow(py).unwrap()
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":442         let array = $array.as_any().downcast_ref::<$ARRAYTYPE>().unwrap();
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":455         let array = $array.as_any().downcast_ref::<$ARRAYTYPE>().unwrap();
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":543                     _ => panic!("Incompatible ScalarValue for list"),
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":566                     _ => panic!("Incompatible ScalarValue for list"),
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":604         let array = $array.as_any().downcast_ref::<$ARRAYTYPE>().unwrap();
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":730             _ => panic!("Cannot run arithmetic negate on scalar value: {:?}", self),
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":892                                 sv => panic!(
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":900                         sv => panic!(
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1095                                 panic!("Expected inner key type of {} but found: {}, value was ({:?})", key_type, inner_key_type, scalar);
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1119                     _ => unreachable!("Invalid dictionary keys type: {:?}", key_type),
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1151                     _ => unreachable!(),
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1165                 _ => unreachable!(),
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1244             .unwrap()
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1370                 .unwrap(),
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1444                     _ => unreachable!("Invalid dictionary keys type: {:?}", key_type),
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1457         let array = array.as_any().downcast_ref::<Decimal128Array>().unwrap();
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1572                     _ => unreachable!("Invalid dictionary keys type: {:?}", key_type),
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1605                     array.as_any().downcast_ref::<FixedSizeListArray>().unwrap();
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1642         let array = array.as_any().downcast_ref::<Decimal128Array>().unwrap();
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1711             ScalarValue::List(_, _) => unimplemented!(),
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1742             ScalarValue::Struct(_, _) => unimplemented!(),
"/home/andy/git/apache/arrow-datafusion/datafusion/common/src/scalar.rs":1753                     _ => unreachable!("Invalid dictionary keys type: {:?}", key_type),
```

**Describe the solution you'd like**
Review code that can panic and see where it makes sense to return a `Result` instead. For example, It is generally better to use `?` than `unwrap`on results.

The goal is not to remove all panics but review and make sure we are using them appropriately. Bonus points for adding documentation for invariants.

**Describe alternatives you've considered**
None

**Additional context**
List generated by https://github.com/andygrove/no-need-to-panic

### PR DESCRIPTION
**Pull request by maruschin**
## Which issue does this PR close?

It's a part of #3313.

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -744,13 +744,13 @@ impl ScalarValue {
     pub fn new_primitive<T: ArrowPrimitiveType>(
         a: Option<T::Native>,
         d: &DataType,
-    ) -> Self {
+    ) -> Result<Self> {
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:747]: I think the comments in this function also need to be updated (the 'panic's section is no longer correct)

```
    /// Create a [`ScalarValue`] with the provided value and datatype
    ///
    /// # Panics
    ///
    /// Panics if d is not compatible with T
```

**Review by alamb — APPROVED** [review]
This looks good to me -- while it is an API change I think it is a reasonable one. Thank you @maruschin 

I noticed the comments also need to be updated but we could also do that as a follow on PR as well

File: datafusion/common/src/scalar.rs (hunk 1 above)
Comment (tustvold) [reply datafusion/common/src/scalar.rs:747]: with_data_type can still panic

Comment (alamb) [reply datafusion/common/src/scalar.rs:747]: Wrong with what? A panic?

Comment (tustvold) [reply datafusion/common/src/scalar.rs:747]: Its an exceptional case that would imply some sort of schema inconsistency within the running plan, so imo panicking is perfectly valid, but then so is returning an error. The broader point is the doc is still correct

Comment (alamb) [reply datafusion/common/src/scalar.rs:747]: I understand the point about the panics -- thank you

I don't have a strong opinion either way with regards to propagating errors vs panic'ing

Comment (maruschin) [reply datafusion/common/src/scalar.rs:755]: @alamb, I rewrote the comment and left a mention about panic. As @tustvold metioned `with_data_type` can panic.

**Review by Dandandan — APPROVED** [review]

**Comment by Dandandan** [comment]
Thanks @maruschin

**Pull request merged**